### PR TITLE
Sales Modular Contract and log

### PR DESCRIPTION
**Pull request by dm-ps**
refs: https://github.com/metasfresh/metasfresh/issues/16177

### CODE REVIEW

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/workpackage/impl/OrderLineBasedModularContractLogsHandler.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+					default -> throw new AdempiereException(MSG_ERROR_DOC_ACTION_UNSUPPORTED);
+				};
+	}
+
+	@Override
+	public @NonNull ExplainedOptional<LogEntryCreateRequest> createLogEntryCreateRequest(@NonNull final CreateLogRequest<I_C_Flatrate_Term> request)
+	{
+		final I_C_Flatrate_Term modularContractRecord = request.getHandleLogsRequest().getModel();
+		final ProductId productId = Optional.ofNullable(ProductId.ofRepoIdOrNull(modularContractRecord.getM_Product_ID()))
+				.orElseThrow(() -> new AdempiereException("Product cannot be null at this point!"));
```
Comment (adi-stefan) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/workpackage/impl/OrderLineBasedModularContractLogsHandler.java:94]:
```suggestion
		final ProductId productId = ProductId.ofRepoId(modularContractRecord.getM_Product_ID());
```

If we're not expecting it to be null, and we're not giving a translated error message, why not use the built in?

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/workpackage/impl/OrderLineBasedModularContractLogsHandler.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+		final OrderId orderId = Optional.ofNullable(OrderId.ofRepoIdOrNull(modularContractRecord.getC_Order_Term_ID()))
+				.orElseThrow(() -> new AdempiereException("OrderTermId cannot be null at this point!"));
```
Comment (adi-stefan) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/workpackage/impl/OrderLineBasedModularContractLogsHandler.java:97]:
```suggestion
		final OrderId orderId = OrderId.ofRepoId(modularContractRecord.getC_Order_Term_ID());
```
same

File: backend/de.metas.contracts/src/main/sql/postgresql/system/50-de.metas.contracts/5701880_sys_gh16177_SalesModularContract.sql
Comment (adi-stefan) [inline backend/de.metas.contracts/src/main/sql/postgresql/system/50-de.metas.contracts/5701880_sys_gh16177_SalesModularContract.sql:1]: Flavouring: I would have expected the file to be either:
- Named differently as to give a hint about its contents or
- Be appended to the file above, if there is no significant difference in terms of contents.

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SOLineForPOModularContractHandler.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+		return logEntryContractType.isModularContractType();
+	}
+
+	@Override
+	public @NonNull Stream<FlatrateTermId> streamContractIds(@NonNull final I_C_OrderLine orderLine)
+	{
+		final I_C_Order order = orderBL.getById(OrderId.ofRepoId(orderLine.getC_Order_ID()));
+
+		final WarehouseId warehouseId = WarehouseId.ofRepoIdOrNull(order.getM_Warehouse_ID());
+		Check.assume(warehouseId != null, "WarehouseId should not be null at this stage!");
```
Comment (adi-stefan) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SOLineForPOModularContractHandler.java:121]:
```suggestion
		final WarehouseId warehouseId = WarehouseId.ofRepoId(order.getM_Warehouse_ID());
```

Same below for harvestingYearId and harvestingCalendarId. No need to use a different null-checking mechanism.

**Review by adi-stefan — COMMENTED** [review]
please see comments.

File: backend/de.metas.contracts/src/main/sql/postgresql/system/50-de.metas.contracts/5701880_sys_gh16177_SalesModularContract.sql
Comment (dm-ps) [reply backend/de.metas.contracts/src/main/sql/postgresql/system/50-de.metas.contracts/5701880_sys_gh16177_SalesModularContract.sql:1]: renamed the file to give more details of what it contains

**Pull request merged**